cryptoxlib/clients/bitpanda/BitpandaWebsocket.py

```diff
@@ -24,12 +25,36 @@ def __init__(self, subscriptions: List[Subscription], api_key: str = None, ssl_c
 
         self.api_key = api_key
 
+    async def _authenticate(self, websocket: Websocket):
+        requires_authentication = False
+        for subscription in self.subscriptions:
+            if subscription.requires_authentication():
+                requires_authentication = True
+                break
+
+        if requires_authentication:
+            authentication_message = {
+                "type": "AUTHENTICATE",
+                "api_token": self.api_key
+            }
+
+            LOG.debug(f"> {authentication_message}")
+            await websocket.send(json.dumps(authentication_message))
+
+            message = await websocket.receive()
+            LOG.debug(f"< {message}")
+
+            message = json.loads(message)
+            if 'type' in message and message['type'] == 'AUTHENTICATED':
+                LOG.info(f"Websocket authenticated successfully.")
+            else:
+                raise BitpandaException(f"Authentication error. Response [{message}]")
```

Comment:
Depending on when this is called, there might be other websocket messages received while authenticating, e.g. if upgrading the connection to an authenticated one, already subscribed (public) channels continue to send messages.

There are also system messages, e.g. the heartbeat that could interfere, maybe these could be ignored until `AUTHENTICATED` is received? https://developers.bitpanda.com/exchange/#idle-connections

Reply:
The way authentication works is that it is the first thing sent in the websocket, subscriptions comes only if the authentication was successful (in case subscriptions require authentication of course). That's the intentional design of the library. So I believe this is okay.

The only potential problem is if heartbeat message is sent between authentication request and authentication response. But 1/ authentication is immediate and 2/ I do not even know if heartbeats can be received for websockets without any subscription. I will test this.